### PR TITLE
ENH PHP 8.1 compatibility

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/10250

### CODE REVIEW

File: src/Control/Controller.php [hunk 1]
```diff
@@ -680,11 +680,11 @@ public static function join_links($arg = null)
         foreach ($args as $arg) {
             // Find fragment identifier - keep the last one
             if (strpos($arg, '#') !== false) {
-                list($arg, $fragmentIdentifier) = explode('#', $arg, 2);
+                list($arg, $fragmentIdentifier) = explode('#', (string) $arg, 2);
```
Comment (michalkleiner) [inline src/Control/Controller.php:683]: Probably needs `$arg = (string) $arg;` as the first line in the loop and then it's all good for all the checks and usages.

File: src/Core/Convert.php [hunk 2]
```diff
@@ -557,22 +557,22 @@ public static function upperCamelToLowerCamel($str)
     {
         $return = null;
         $matches = null;
-        if (preg_match('/(^[A-Z]{1,})([A-Z]{1})([a-z]+.*)/', $str, $matches)) {
+        if (preg_match('/(^[A-Z]{1,})([A-Z]{1})([a-z]+.*)/', (string) $str, $matches)) {
```
Comment (michalkleiner) [inline src/Core/Convert.php:560]: Retype as the first thing in the method and no need to worry about it in 3 different places afterwards?

Comment (emteknetnz) [reply src/Core/Convert.php:560]: This was the largely the result of running rector/rector over the codebase, which will simply cast things to `(string)` to ensure they are not `null`.

This work looks to be bigger than originally envisioned (needs to be done across all of silverstripe/recipe-kitchen-sink, or not at all), so we're currently weighing up the priority of this against other competing priorities

**Comment by GuySartorelli** [comment]
The error in the travis build is unrelated and is fixed by silverstripe/silverstripe-travis-shared#36

**Review by GuySartorelli — APPROVED** [review]
For posterity: See discussion in silverstripe/silverstripe-admin#1294 about the general approach. tl;dr:
> the purposes of this PR is to avoid passing null (specifically null) from built-in functions that as of php 8.1 will throw deprecation warnings if null is passed in. The approach is necessarily heavy-handed, and while there are many situations where it isn't needed, it would be prohibitively laborious to find all of those situations by hand.

Additionally, where `#[\ReturnTypeWillChange]` has been used, this is to avoid a breaking change in our API where php8.1 now enforces a given return type.

Looks good to me

**Pull request merged**